File: 力扣/学习计划/剑指 Offer I/008_数组中重复的数字/main.go
package main

import "fmt"

// https://leetcode-cn.com/problems/shu-zu-zhong-zhong-fu-de-shu-zi-lcof/

// 迭代 + map
func findRepeatNumber(nums []int) int {
	m := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		if val, ok := m[nums[i]]; ok {
			return val
		}
		m[nums[i]] = nums[i]
	}
	return 0
}

// 迭代 + 切片
func findRepeatNumber2(nums []int) int {
	records := make([]int, len(nums))
	for i := 0; i < len(nums); i++ {
		records[nums[i]]++
		if records[nums[i]] > 1 {
			return nums[i]
		}
	}
	return 0
}

// 题解：原地交换
func findRepeatNumber3(nums []int) int {
	i := 0
	for i < len(nums) {
		if nums[i] == nums[nums[i]] {
			i++
			continue
		}
		tmp := nums[i]
		nums[i] = nums[nums[i+1]]
		nums[nums[i+1]] = tmp
	}
	return -1
}

func main() {
	// findRepeatNumber(nil)
	m := make(map[string]interface{})
	m["a"] = "AAA"
	if _, ok := m["ba"]; ok {
		fmt.Println("存在")
	} else {
		fmt.Println("不存在")
	}
}

File: 力扣/学习计划/剑指 Offer I/008_数组中重复的数字/main.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

// https://leetcode-cn.com/problems/shu-zu-zhong-zhong-fu-de-shu-zi-lcof/


class Solution {
    /**
     * 迭代 + map 
     */
    public int findRepeatNumber(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i])) {
                return nums[i];
            }
            map.put(nums[i], nums[i]);
        }
        return 0;
    }
    
    /**
     * 迭代 + 数组
     */
    public int findRepeatNumber2(int[] nums) {
        int[] records = new int[nums.length];
        for (int i = 0; i < nums.length; i++) {
            records[nums[i]]++;
            if (records[nums[i]] > 1) {
                return nums[i];
            }
        }
        return 0;
    }
    /**
     * 题解：迭代 + set
     */
    public int findRepeatNumber3(int[] nums) {
        Set<Integer> dic = new HashSet<>();
        for (int num : nums) {
            if (dic.contains(num))
                return num;
            dic.add(num);
        }
        return 0;
    }
}